src/service/Fetch.ts

```diff
@@ -1,63 +1,105 @@
+import LAMP from '../index'
+import { Semaphore } from "await-semaphore"
 /**
  *
  */
-export type Configuration = {
-  /**
-   *
-   */
-  base: string
-
-  /**
-   *
-   */
-  authorization?: string
-
-  /**
-   *
-   */
-  headers?: { [header: string]: string }
+
+ interface UpdateTokenResult {
+  access_token: string
+  refresh_token: string
+  success: boolean
 }
 
-async function _fetch<ResultType>(
+const _tokenSemaphore = new Semaphore(1)
+
+const _fetch = async <ResultType>(
   method: string,
   route: string,
-  configuration?: Configuration,
-  body?: any
-): Promise<ResultType> {
-  if (!configuration) throw new Error("Cannot make HTTP request due to invalid configuration.")
+  body?: any,
+  tryRefreshToken: boolean = true,
+  waitForToken: boolean = true,
+): Promise<ResultType> => {
+  if (waitForToken) {
+    const release = await _tokenSemaphore.acquire()
+    release()
+  }
 
-  var result = await (
-    await fetch(`${configuration!.base}${route}`, {
+  if (!LAMP.Auth._auth)
+    throw new Error("Cannot make HTTP request due to invalid configuration.")
+
+  const { id, password, accessToken } = LAMP.Auth._auth
+  const authorization = accessToken
+    ? `Bearer ${accessToken}`
+    : id && password
+    ? `Basic ${id}:${password}`
+    : undefined
+
+  const response = await fetch(
+    `https://${LAMP.Auth._auth.serverAddress}${route}`,
+    {
       method: method,
       headers: new Headers({
         "Content-Type": "application/json",
         Accept: "application/json",
-        ...(configuration!.headers || {}),
-        Authorization: !!configuration!.authorization ? `Basic ${configuration!.authorization}` : undefined
+        Authorization: authorization
       } as any),
-      body: body !== undefined ? JSON.stringify(body) : undefined
-    })
-  ).json()
+      body: !!body ? JSON.stringify(body) : undefined
+    }
+  )
 
-  //if (result['error'] !== undefined)
-  //    throw new Error(result['error'])
-  return result as any
+  if (response.status === 401 && tryRefreshToken) { // The access token has expired; try to redeem refresh token
+    await refreshToken(accessToken)
+    return _fetch(method, route, body, false)
+  }
+
+  return response.json()
 }
 
+const refreshToken = async (expiredToken: string) =>
```

Comment:
It *feels* to me like the token refresh is automated where users will never have an automatic logout issue (i.e. once they manually log in, they should stay logged in until manually logging out), but is that going to be the case?

Reply:
We should switch back to localStorage to avoid this issue